firedrake/mesh.py

```diff
@@ -93,23 +88,33 @@ def _null_subset(self):
 
         return op2.Subset(self.set, [])
 
-    def measure_set(self, integral_type, subdomain_id):
+    def measure_set(self, integral_type, subdomain_id,
+                    all_integer_subdomain_ids=()):
         '''Return the iteration set appropriate to measure. This will
         either be for all the interior or exterior (as appropriate)
         facets, or for a particular numbered subdomain.'''
 
-        # ufl.Measure doesn't have enums for these any more :(
-        # FIXME: This is WRONG in the case of 1*(ds + ds(1))
-        # "otherwise" is "everywhere" - set(explicit subdomain ids)
-        if subdomain_id in ["everywhere", "otherwise"]:
-            if integral_type == "exterior_facet_bottom":
-                return (op2.ON_BOTTOM, self.bottom_set)
-            elif integral_type == "exterior_facet_top":
-                return (op2.ON_TOP, self.bottom_set)
-            elif integral_type == "interior_facet_horiz":
-                return self.bottom_set
-            else:
+        if integral_type in ("exterior_facet_bottom",
+                             "exterior_facet_top",
+                             "interior_facet_horiz"):
+            # these iterate over the base cell set
+            return self.mesh.cell_subset(subdomain_id, all_integer_subdomain_ids)
+
+        if subdomain_id == "everywhere":
+            return self.set
+        if subdomain_id == "otherwise":
+            if len(all_integer_subdomain_ids) == 0:
                 return self.set
+            key = ("otherwise", ) + all_integer_subdomain_ids
+            try:
+                return self._subsets[key]
+            except KeyError:
+                ids = [np.where(self.markers == sid)[0]
+                       for sid in all_integer_subdomain_ids]
+                to_remove = np.unique(np.concatenate(ids))
+                indices = np.arange(self.set.total_size, dtype=np.int32)
+                indices = np.delete(indices, to_remove)
```

Comment:
Some testing suggests that the two are the same complexity, so this is clearer, I think.